File: app/src/main/java/com/callmangement/EHR/support/CompressImage.java
package com.callmangement.EHR.support;

import android.annotation.SuppressLint;
import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.media.ExifInterface;
import android.net.Uri;
import android.provider.MediaStore;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class CompressImage {

    public static String compress(String mfilepath, Context context) {

        try {

            Bitmap scaledBitmap = null;

            BitmapFactory.Options options = new BitmapFactory.Options();

            options.inJustDecodeBounds = true;
            Bitmap bmp = BitmapFactory.decodeFile(mfilepath, options);

            int actualHeight = options.outHeight;
            int actualWidth = options.outWidth;

            /*float maxHeight = 816.0f;
            float maxWidth = 612.0f;*/
            /*float maxHeight = 1920.0f;
            float maxWidth = 1080.0f;*/

            float maxHeight = 1200.0f;
            float maxWidth = 944.0f;
            float imgRatio = actualWidth / actualHeight;
            float maxRatio = maxWidth / maxHeight;

            if (actualHeight > maxHeight || actualWidth > maxWidth) {
                if (imgRatio < maxRatio) {
                    imgRatio = maxHeight / actualHeight;
                    actualWidth = (int) (imgRatio * actualWidth);
                    actualHeight = (int) maxHeight;
                } else if (imgRatio > maxRatio) {
                    imgRatio = maxWidth / actualWidth;
                    actualHeight = (int) (imgRatio * actualHeight);
                    actualWidth = (int) maxWidth;
                } else {
                    actualHeight = (int) maxHeight;
                    actualWidth = (int) maxWidth;

                }
            }

            //      setting inSampleSize value allows to load a scaled down version of the original image

            options.inSampleSize = calculateInSampleSize(options, actualWidth, actualHeight);

            //      inJustDecodeBounds set to false to load the actual bitmap
            options.inJustDecodeBounds = false;

            //      this options allow android to claim the bitmap memory if it runs low on memory
            options.inPurgeable = true;
            options.inInputShareable = true;
            options.inTempStorage = new byte[16 * 1024];

            try {
                //          load the bitmap from its path
                bmp = BitmapFactory.decodeFile(mfilepath, options);
            } catch (OutOfMemoryError exception) {
                exception.printStackTrace();

            }
            try {
                scaledBitmap = Bitmap.createBitmap(actualWidth, actualHeight, Bitmap.Config.ARGB_8888);
            } catch (OutOfMemoryError exception) {
                exception.printStackTrace();
            }

            float ratioX = actualWidth / (float) options.outWidth;
            float ratioY = actualHeight / (float) options.outHeight;
            float middleX = actualWidth / 2.0f;
            float middleY = actualHeight / 2.0f;

            Matrix scaleMatrix = new Matrix();
            scaleMatrix.setScale(ratioX, ratioY, middleX, middleY);

            Canvas canvas = new Canvas(scaledBitmap);
            canvas.setMatrix(scaleMatrix);
            canvas.drawBitmap(bmp, middleX - bmp.getWidth() / 2, middleY - bmp.getHeight() / 2, new Paint(Paint.FILTER_BITMAP_FLAG));

            //      check the rotation of the image and display it properly
            ExifInterface exif;
            try {
                exif = new ExifInterface(mfilepath);

                int orientation = exif.getAttributeInt(
                        ExifInterface.TAG_ORIENTATION, 0);

                Matrix matrix = new Matrix();
                if (orientation == 6) {
                    matrix.postRotate(90);

                } else if (orientation == 3) {
                    matrix.postRotate(180);

                } else if (orientation == 8) {
                    matrix.postRotate(270);

                }
                scaledBitmap = Bitmap.createBitmap(scaledBitmap, 0, 0,
                        scaledBitmap.getWidth(), scaledBitmap.getHeight(), matrix,
                        true);
            } catch (IOException e) {
                e.printStackTrace();
            }

            FileOutputStream out;
            String filepath = getFilename(context);
            try {
                out = new FileOutputStream(filepath);
                //          write the compressed bitmap at the destination specified by filename.
                scaledBitmap.compress(Bitmap.CompressFormat.JPEG, 100, out);
                return filepath;
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
        return "";
    }

    private static int calculateInSampleSize(BitmapFactory.Options options, int reqWidth, int reqHeight) {
        final int height = options.outHeight;
        final int width = options.outWidth;
        int inSampleSize = 1;

        if (height > reqHeight || width > reqWidth) {
            final int heightRatio = Math.round((float) height / (float) reqHeight);
            final int widthRatio = Math.round((float) width / (float) reqWidth);
            inSampleSize = heightRatio < widthRatio ? heightRatio : widthRatio;
        }
        final float totalPixels = width * height;
        final float totalReqPixelsCap = reqWidth * reqHeight * 2;
        while (totalPixels / (inSampleSize * inSampleSize) > totalReqPixelsCap) {
            inSampleSize++;
        }

        return inSampleSize;
    }

    public static String getFilename(Context context) {

        File file = new File(context.getFilesDir()/*Environment.getExternalStorageDirectory().getPath()*/, "Epdsfrt/Images");
        if (!file.exists()) {
            file.mkdirs();
        }
        String uriSting = (file.getAbsolutePath() + "/" + System.currentTimeMillis() + ".jpg");
        return uriSting;

    }

    private String getRealPathFromURI(String contentURI, Context context) {
        Uri contentUri = Uri.parse(contentURI);
        @SuppressLint("Recycle") Cursor cursor = context.getContentResolver().query(contentUri, null, null, null, null);
        if (cursor == null) {
            return contentUri.getPath();
        } else {
            cursor.moveToFirst();
            int index = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA);
            return cursor.getString(index);
        }
    }

}

File: app/src/main/java/com/callmangement/utils/CompressImage.kt
package com.callmangement.utils

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.media.ExifInterface
import android.net.Uri
import android.provider.MediaStore
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException

class CompressImage {
    private fun getRealPathFromURI(contentURI: String, context: Context): String? {
        val contentUri = Uri.parse(contentURI)
        @SuppressLint("Recycle") val cursor =
            context.contentResolver.query(contentUri, null, null, null, null)
        if (cursor == null) {
            return contentUri.path
        } else {
            cursor.moveToFirst()
            val index = cursor.getColumnIndex(MediaStore.Images.ImageColumns.DATA)
            return cursor.getString(index)
        }
    }

    companion object {
        fun compress(mfilepath: String, context: Context): String {
            try {
                var scaledBitmap: Bitmap? = null

                val options = BitmapFactory.Options()

                options.inJustDecodeBounds = true
                var bmp = BitmapFactory.decodeFile(mfilepath, options)

                var actualHeight = options.outHeight
                var actualWidth = options.outWidth

                /*float maxHeight = 816.0f;
            float maxWidth = 612.0f;*/
                /*float maxHeight = 1920.0f;
            float maxWidth = 1080.0f;*/
                val maxHeight = 1200.0f
                val maxWidth = 944.0f
                var imgRatio = (actualWidth / actualHeight).toFloat()
                val maxRatio = maxWidth / maxHeight

                if (actualHeight > maxHeight || actualWidth > maxWidth) {
                    if (imgRatio < maxRatio) {
                        imgRatio = maxHeight / actualHeight
                        actualWidth = (imgRatio * actualWidth).toInt()
                        actualHeight = maxHeight.toInt()
                    } else if (imgRatio > maxRatio) {
                        imgRatio = maxWidth / actualWidth
                        actualHeight = (imgRatio * actualHeight).toInt()
                        actualWidth = maxWidth.toInt()
                    } else {
                        actualHeight = maxHeight.toInt()
                        actualWidth = maxWidth.toInt()
                    }
                }

                //      setting inSampleSize value allows to load a scaled down version of the original image
                options.inSampleSize = calculateInSampleSize(options, actualWidth, actualHeight)

                //      inJustDecodeBounds set to false to load the actual bitmap
                options.inJustDecodeBounds = false

                //      this options allow android to claim the bitmap memory if it runs low on memory
                options.inPurgeable = true
                options.inInputShareable = true
                options.inTempStorage = ByteArray(16 * 1024)

                try {
                    //          load the bitmap from its path
                    bmp = BitmapFactory.decodeFile(mfilepath, options)
                } catch (exception: OutOfMemoryError) {
                    exception.printStackTrace()
                }
                try {
                    scaledBitmap =
                        Bitmap.createBitmap(actualWidth, actualHeight, Bitmap.Config.ARGB_8888)
                } catch (exception: OutOfMemoryError) {
                    exception.printStackTrace()
                }

                val ratioX = actualWidth / options.outWidth.toFloat()
                val ratioY = actualHeight / options.outHeight.toFloat()
                val middleX = actualWidth / 2.0f
                val middleY = actualHeight / 2.0f

                val scaleMatrix = Matrix()
                scaleMatrix.setScale(ratioX, ratioY, middleX, middleY)

                val canvas = Canvas(scaledBitmap!!)
                canvas.setMatrix(scaleMatrix)
                canvas.drawBitmap(
                    bmp, middleX - bmp.width / 2, middleY - bmp.height / 2, Paint(
                        Paint.FILTER_BITMAP_FLAG
                    )
                )

                //      check the rotation of the image and display it properly
                val exif: ExifInterface
                try {
                    exif = ExifInterface(mfilepath)

                    val orientation = exif.getAttributeInt(
                        ExifInterface.TAG_ORIENTATION, 0
                    )

                    val matrix = Matrix()
                    if (orientation == 6) {
                        matrix.postRotate(90f)
                    } else if (orientation == 3) {
                        matrix.postRotate(180f)
                    } else if (orientation == 8) {
                        matrix.postRotate(270f)
                    }
                    scaledBitmap = Bitmap.createBitmap(
                        scaledBitmap, 0, 0,
                        scaledBitmap.width, scaledBitmap.height, matrix,
                        true
                    )
                } catch (e: IOException) {
                    e.printStackTrace()
                }

                val out: FileOutputStream
                val filepath = getFilename(context)
                try {
                    out = FileOutputStream(filepath)
                    //          write the compressed bitmap at the destination specified by filename.
                    scaledBitmap!!.compress(Bitmap.CompressFormat.JPEG, 100, out)
                    return filepath
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return ""
        }

        private fun calculateInSampleSize(
            options: BitmapFactory.Options,
            reqWidth: Int,
            reqHeight: Int
        ): Int {
            val height = options.outHeight
            val width = options.outWidth
            var inSampleSize = 1

            if (height > reqHeight || width > reqWidth) {
                val heightRatio = Math.round(height.toFloat() / reqHeight.toFloat())
                val widthRatio = Math.round(width.toFloat() / reqWidth.toFloat())
                inSampleSize = if (heightRatio < widthRatio) heightRatio else widthRatio
            }
            val totalPixels = (width * height).toFloat()
            val totalReqPixelsCap = (reqWidth * reqHeight * 2).toFloat()
            while (totalPixels / (inSampleSize * inSampleSize) > totalReqPixelsCap) {
                inSampleSize++
            }

            return inSampleSize
        }

        fun getFilename(context: Context): String {
            val file = File(
                context.filesDir,  /*Environment.getExternalStorageDirectory().getPath()*/
                "Epdsfrt/Images"
            )
            if (!file.exists()) {
                file.mkdirs()
            }
            val uriSting = (file.absolutePath + "/" + System.currentTimeMillis() + ".jpg")
            return uriSting
        }
    }
}
